Попов Захар

BlizPerfect opened the pull request:
@Folleach

Проведено начальное проектирование архитектуры программы.

Идея:
Сначала входная строка разбивается на упорядоченный список токенов. Токен - сущность, хранящая в себе часть строки и соответствующий этой части тег.
Есть три типа токенов:
1. Простой токен. Хранит в себе строку без какого-либо форматирования.
2. Токен с разметкой. Хранит в себе строку с форматированием без вложений других токенов.
3. Составной токен. Хранит в себе строку с форматированием и вложением других токенов.

Затем, в списке токенов подменяются теги из markdown разметки на теги HTML разметки и список токенов собирается обратно в строку. В Program.cs привёл 4 примера токенов для разных ситуаций, что бы было наглядней. 

Пара мыслей:
1. Хочется отдельно сказать про переопределённый в ComplexSringToken метод ToString().
Не отпускает идея того, что бы токен в ToString() сам возвращал своё строковое представление. Я понимаю, что это как раз работа Writer'a, но я сам как-то зациклился на этой идее. В черновой реализации каждый раз, когда составной токен возвращает своё строковое представление, создаётся новый экземпляр StringBuilder(), что расточительно. Может быть просто использовать один статический StringBuilder() для всех ToString() в ComplexSringToken, который по концу ToString() будет очищаться?
2. Хочется услышать мнение о целесообразности приведения конкретных реализаций парсера, конвертера и врайтера в статические, а не динамические классы. Тут больше вопрос в том, как их тогда обязать реализовывать интерфейсы?
3. При написании черновой архитектуры попробовал сделать особый упор на минимизацию действий при создании новых тегов. Например, для создания заголовков второго уровня (\#\#), не потребуется создавать новый кодовый файл.
4. Помню про включение опции невозможности компиляции программы при предупреждениях, подключу её уже при написании внутренностей методов.

BlizPerfect committed:
Проведено начальное проектирование.

Folleach commented on cs/Markdown/Tags/Tag.cs:
```diff
@@ -0,0 +1,16 @@
+            TagName = tagName;
+        }
+        public string TagName { get; }
+    }
```
Наличие класса `Tag` вместе с интерфейсом `ITag` несколько запутывает.  
Я понял, что `Tag` обязует указать имя тега, который будет распознан как начало токена, причём проверяет, что `tagName` не пустой.  
В то же время `ITag` обязует указать открывающий и закрывающий тег.
И применяются они в паре
```cs
public class SomeTag : Tag, ITag
```

В таком случае задаёшься вопросом, почему тогда сам `Tag` не реализует `ITag`?
```cs
internal abstract class Tag : ITag
```

Но поглядев на их ответственность, кажется, она разная? `Tag` это про парсинг исходного markdown текста, `ITag` это про рендер текста в другой формат (html в нашем случае).

В таком случае одинаковое название всё ещё запутывает. Оба теги, но первый это одно, второй это совершенно другое

Folleach replied:
resolved

Folleach commented on cs/Markdown/Converters/MdToHTMLConverters/ConverterMdToHTML.cs:
```diff
@@ -0,0 +1,22 @@
+        public IList<Token> Convert(IList<Token> tokensToConvert)
+        {
+            throw new NotImplementedException();
+        }
```
А, даже так, `Token` сначала распознаётся из `Markdown`, в токене сохраняется разметка, которая потом будет преобразована в такой же токен но с другой разметкой?  
В таком случае `Tag` и `ITag` действительно становятся чуточку похожи, но тогда другой вопрос:
Точно ли хранение визуального представления должно быть ответственностью токена?  

Вероятно, хотелось бы абстрагироваться от визуальных образов того или иного текста.  
Во первых визуальный образ токена может быть сложным, например в максимальном решении теги имеют не только парный закрывающий и открывающий тег, но и другую информацию.  
```markdown
[link name](https://example.org/)
```
Во вторых, потенциально мы могли бы захотеть рендерить `Markdown` там, где неприменима html разметка. Ну, например, отдавая команды напрямую в OpenGL.  В таком случае не понятно, что должен хранить токен.

В мире подобная проблема уже решена с помощью [абстрактного синтаксического дерева (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree)   
Если прочитаешь его мотивацию, то, вероятнее всего, будут видны некоторые сходства с этой задачей.  

Не говорю, что именно тут нужно реализовать AST в полной мере, но идеи оттуда взять можно.

Folleach replied:
Что-ж, тегов больше не существует, хех.  
Мне понравилось, как токены превратились в абстрактные сущности, которые содержат только нужную им информацию. Прямо круто!

resolved

Folleach commented on cs/Markdown/Tags/MdTags/SingleMdTag.cs:
```diff
@@ -0,0 +1,9 @@
+        public string OpeningTagName => TagName;
+
+        public string ClosingTagName => string.Empty;
+    }
```
Не нарушается ли тут Interface Segregation Principle?
https://en.wikipedia.org/wiki/SOLID

Конкретно в свойстве `ClosingTagName`

Folleach replied:
Ответ был в личных сообщениях.

resolved

Folleach commented on cs/Markdown/Md.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Markdown
+{
+    internal class Md(IParser parser, IConverter converter, IWriter writer)
```
Это хорошо, что класс `Md` конфигурируется.  
Но некоторые части лучше реализовать внутри. Например, точно ли класс для markdown должен обязывать пользователя создавать для него правила?

Нужно оставить только минимально необходимый набор параметров, которые должны быть получены извне.

Folleach replied:
Могу ли я создать класс `Md` и передать в него не `Markdown` парсер?  
![image](https://github.com/user-attachments/assets/30ec97df-1661-463e-9d61-235a2b3081e2)

Получается класс `Md` это какая-то абстрактная машина конвертации форматов? Правильное ли тогда название `Md`?

Folleach commented on cs/Markdown/Md.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        private readonly IParser _parser = parser;
+        private readonly IConverter _converter = converter;
+        private readonly IWriter _writer = writer;
```
Сейчас в классе есть такие поля:
- parser
- _parser
- converter
- _converter
- writer
- _writer

Нужно оставить только один набор )
Это потому, что параметры primary constructor видны во всём классе
https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/primary-constructors#primary-constructors
> Primary constructor parameters are in scope throughout the class definition. It's important to view primary constructor parameters as parameters even though they are in scope throughout the class definition. Several rules clarify that they're parameters

Folleach replied:
resolved

Folleach commented on cs/Markdown/Md.cs:
```diff
@@ -0,0 +1,22 @@
+        {
+            var mdTokens = _parser.ParseToTokens(text);
+
+            var HTMLTokens = _converter.Convert(mdTokens);
```
Локальные переменные принято называть с маленькой буквы.  
Ничего страшного, если HTML будет написан в нижнем регистре: `htmlTokens`  
То же самое касается и названий классов. Думаю название `HtmlTag` - нормальное

Folleach replied:
resolved

Folleach commented on cs/Markdown/Program.cs:
```diff
@@ -0,0 +1,82 @@
+            var simpleText = "Hello, Andrey";
+            var simpleTokenMD = new SimpleStringToken(simpleText);
+            var simpleTokenHTML = new SimpleStringToken("Hello, Andrey");
+            ShowEaxmple(simpleText, simpleTokenMD, simpleTokenHTML);
```
Можно было вполне реализовать сразу тестами :)
Но в целом понятно почему тут так сделано, никакой логики не тестируется, код служит для демонстрации.  

После прочтения можешь сразу закрывать этот комментарий

Folleach replied:
resolved

Folleach commented on cs/Markdown/Program.cs:
```diff
@@ -0,0 +1,82 @@
+{
+    internal class Program
+    {
+        static void Main(string[] args)
```
Правда ли этот проект должен быть исполняемым файлом?  
Давай сделаем его библиотекой

Folleach replied:
Можно даже удалить `OutputType` параметр. Он по-умолчанию `Library`
https://learn.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-properties?view=vs-2022

resolved

Folleach commented on cs/Markdown/Tokens/MarkdownStringToken.cs:
```diff
@@ -0,0 +1,28 @@
+    {
+        public ITag Tag { get => _tag!; }
+
+        public string Content { get => _content!; }
```
Есть более простой синтаксис написать read-only свойство
https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members#read-only-properties

Folleach replied:
resolved

Folleach commented on cs/Markdown/Tokens/ComplexSringToken.cs:
```diff
@@ -0,0 +1,50 @@
+            sb.Append(Tag.ClosingTagName);
+
+            return sb.ToString();
+        }
```
Прокомментирую тут первую мысль из описания к PR
> Хочется отдельно сказать про переопределённый в ComplexSringToken метод ToString().

Это действитеьно больше похоже на работу `IWriter`, можно подумать в сторону: "а что, если мы захотим добавить рендер, отличный от html?".  
В таком случае придётся переписать каждый токен. А в чём будет заключаться тогда работа `IWriter`?
```cs
return string.Join(string.Empty, tokensToWrite.Select(token => token.ToString());
```

Выглядит странно )
Поэтому точно нужно вынести эту логику из токена.

> В черновой реализации каждый раз, когда составной токен возвращает своё строковое представление, создаётся новый экземпляр StringBuilder()

Ну, в какой-то степени это можно назвать "расточительство". Для решения этой проблемы можно использовать какой-то из паттернов проектирования. Какой бы ты применил?
https://en.wikipedia.org/wiki/Software_design_pattern

Статический `StringBuilder` точно использовать не стоит, иначе может возникнуть ситуация когда два разных экземпляра класса одновременно используют этот `StringBuilder`, тем самым испортят себе состояния

Folleach replied:
Да, паттер Visitor тут хорошее решение.  

resolved

Folleach commented on cs/Markdown/Writers/IWriter.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    internal interface IWriter
+    {
+        string Write(IList<Token> tokensToWrite);
```
Суффикс, кажется, перегружен.  
Если читать как текст, то выходит:
```
writer.write(tokens to write)
```

Folleach replied:
resolved

Folleach commented on cs/Markdown/Converters/IConverter.cs:
```diff
@@ -0,0 +1,9 @@
+    internal interface IConverter
+    {
+        IList<Token> Convert(IList<Token> tokensToConvert);
+    }
```
Добавлю тут про второй пункт из описания.

> Хочется услышать мнение о целесообразности приведения конкретных реализаций парсера, конвертера и врайтера в статические, а не динамические классы. Тут больше вопрос в том, как их тогда обязать реализовывать интерфейсы?

Если реализация у метода/класса одна, и в будущем не предвидится появление новых реализаций - то это можно реализовать статическим классом без дополнительного интерфейса.
Если же реализаций точно будет несколько, то в таком случае статика подойдёт тут в меньшей степени.
Нормально сделать интерфейс, который объявляет какие-либо методы и иметь динамические классы для него.

Что касается конкретно `IConverter`, `IWriter` и `IParser`, то наверное вопрос к тебе, что по твоему мнению никогда не потребует дополнительной реализации? А что может потребовать?

Folleach commented on cs/Markdown/Parsers/MdParsers/ParserMD.cs:
```diff
@@ -0,0 +1,21 @@
+            {"_", new PairedMdTag("_") },
+            {"__", new PairedMdTag("__") },
+            {"# ", new SingleMdTag("# ") }
+        };
```
```cs
private readonly Dictionary<string, ITag> _mdTags = new Dictionary<string, ITag>()
```

Можно сократить до
```cs
private readonly Dictionary<string, ITag> _mdTags = new()
```

Folleach replied:
resolved

Folleach commented on cs/Markdown/Tokens/ComplexSringToken.cs:
```diff
@@ -0,0 +1,50 @@
+            {
+                if (token is SimpleStringToken || token is MarkdownStringToken)
+                {
+                    sb.Append(token.ToString());
```
`ToString` будет вызван неявно, можно его опустить
```cs
sb.Append(token);
```

Folleach replied:
resolved

Folleach reviewed (commented):
Итого
Здорово, что теперь ветка заведена специально под задачу!

Сейчас нужно чуть улучшить декомпозицию, конкретно к сущности "Токен" у меня есть некоторые вопросы, которые написаны ниже.
А также реализовать оставшиеся методы и написать к ним тесты. То есть довести задачу до её финального вида и не забыть про предупреждения от IDE, хех :)

Кстати, некоторые комментарии ниже хоть и оставлены к конкретным строчкам, но от них ожидаются исправления во всём коде.

BlizPerfect committed:
В работе приложения больше не используются конвёртеры тегов.

BlizPerfect committed:
В работе приложения больше не используются теги в явном виде.

BlizPerfect committed:
Реализация с помощью паттерна "Посетитель" получения визуального представдения токена.

BlizPerfect committed:
Рефакторинг токенов.

BlizPerfect committed:
Реализация парсера.

BlizPerfect committed:
Смена типа выходного файла на библиотеку.

BlizPerfect committed:
Написание тестов

BlizPerfect committed:
Исправил один int в цикле на var.

BlizPerfect committed:
Выполнен рефакторинг тестов и изменено поведение парсера.

1. Выполнен рефакторинг тестов. В связи с тем, что одно и тоже визуальное представление переданного текста можно получить из разных комбинаций токенов, я решил оставить в ParserMdTests только тривиальные ситуации, а в MdTests перенести все тесты, какие были раньше, но проверять именно корректность получения визуального представления. В связи с этим из ParserMd был убран избыточный функционал.

BlizPerfect committed:
Рефакторинг тестов в MdTests.

BlizPerfect committed:
Рефакторинг ParserMd

Folleach commented on cs/Markdown.Tests/ParsersTests/ParserMdTest.cs:
```diff
@@ -0,0 +1,222 @@
+            _expected.Clear();
+        }
+
+        [TestCase("__Bold token__")]
+        public void Parse_SimpleBoldText_Correctly(string text)
+        {
+            _expected.Add(
+                new BoldToken(
+                    new TextToken("Bold token")));
+            CheckCorrectness(text);
+        }
```
![image](https://github.com/user-attachments/assets/52fd0674-98f3-42fd-8c68-ec3e53d727fa)

Эти тесты точно корректны?

Folleach replied:
Тесты из `ParserMdTests` не проверяют правильность созданных токенов из текста.
Тесты из `RendererHtmlTests` рендерят html из заданных токенов.

Следовательно, я не могу верить, что текст `_hello_` будет отредерен как `<em>hello</em>`

Folleach replied:
На сколько я понял, проблема была решена добавлением поля `Type` в токены.
На это поле никто не ссылается, всё что оно делает, это занимает 4 байта в каждом токене только для того, чтобы работали тесты? Это странно.
Код не должен содержать специальных для тестов объектов.

Возможно можно было сравнить альтернативным способом.
Быстрый взгляд на документацию не подсказал удобного способа сравнить типы. Поэтому, вероятно, код был бы сложнее.

Более простым вариантом можно было бы добавить в `BasicMarkdownToken` свойство:
```cs
public Type TokenType => this.GetType();
```

Всё ещё костыль, но хотя бы имеем меньше шансов забыть указать правильный TokenType при написании нового токена.
А также не занимает 4 байта в экземпляре токена.

resolve

Folleach commented on cs/Markdown.Tests/ParsersTests/ParserMdTest.cs:
```diff
@@ -0,0 +1,222 @@
+            CheckCorrectness(text);
+        }
+
+        [TestCase(@"\\_Italic token\\_")]
+        public void Parse_ItalicTokenEscapedEscapeSymbol_Correctly(string text)
+        {
+            _expected.Add(
+                new ItalicToken(
+                    new TextToken("Italic token")));
+            CheckCorrectness(text);
+        }
```
Но символ `\` должен сохраниться, его ведь экранировали для отображения.

Так: "\\_Italic token\\_" это выглядит в markdown на github

Из [спецификации](https://github.com/kontur-courses/clean-code/blob/master/MarkdownSpec.md):
> Символ экранирования тоже можно экранировать

Folleach replied:
resolved

Folleach commented on cs/Markdown.Tests/ParsersTests/ParserMdTest.cs:
```diff
@@ -0,0 +1,222 @@
+            CheckCorrectness(text);
+        }
+
+        [TestCase("__Outer bold _Inner italic part_ outer bold__")]
+        public void Parse_ItalicInsideBold_Correctly(string text)
+        {
+            _expected.Add(
+                new BoldToken(
+                    new SetToken(
+                        new TextToken("Outer bold "),
+                        new ItalicToken(
+                            new TextToken("Inner italic part")),
+                        new TextToken(" outer bold"))));
+            CheckCorrectness(text);
+        }
```
Кажется, в тестах не хватает случая наоборот

Folleach replied:
А ещё не вижу случая с `цифрами_1_23`

Folleach replied:
Ещё случай с `__непарными_` тегами

Folleach replied:
> За подчерками, начинающими выделение, должен следовать непробельный символ. Иначе эти_ подчерки_ не считаются выделением и остаются просто символами подчерка.

Тоже не вижу теста, и наоборот

Folleach replied:
> Если внутри подчерков пустая строка ____, то они остаются символами подчерка.

Folleach replied:
До меня дошло, огромный словарь из `MdTests` используется не только как шаблон для нагрузочного теста.  

В таком случае все случаи из этого комментариях покрыты.
resolved

Folleach commented on cs/Markdown.Tests/ParsersTests/ParserMdTest.cs:
```diff
@@ -0,0 +1,222 @@
+namespace Markdown.Tests.ParsersTests
+{
+    [TestFixture]
+    public class ParserMdTests
```
Напишу сюда, так как теста на случай с пересечением тегов нету
> В случае `__пересечения _двойных__ и одинарных_` подчерков ни один из них не считается выделением.

Сейчас выделением считается первое попадание. Что нарушает спецификацию.
![image](https://github.com/user-attachments/assets/be8ca8e1-7562-4ef3-8f04-5ddc9851813f)

Folleach replied:
Вроде проблема осталась
![image](https://github.com/user-attachments/assets/6e7456ac-c9ef-45d7-bd85-743a5b725f01)

Тест `__Hello_ __world_` не совсем релевантен, потому что по спецификации `__hello __` не может быть превращён в `<strong>hello </strong>` из-за пробела.  
> За подчерками, начинающими выделение, должен следовать непробельный символ. Иначе эти_ подчерки_ не считаются выделением и остаются просто символами подчерка.

Folleach commented on cs/Markdown.Tests/ParsersTests/ParserMdTest.cs:
```diff
@@ -0,0 +1,222 @@
+namespace Markdown.Tests.ParsersTests
+{
+    [TestFixture]
+    public class ParserMdTests
```
Хорошо, что тесты на заголовок есть, но нету случая с тем, когда заголовок начинается не в начале абзаца.

Из спецификации:
> **Абзац, начинающийся с "# "**, выделяется тегом \<h1> в заголовок.

![image](https://github.com/user-attachments/assets/8472250f-16b3-4c58-8b62-0ff405b17966)

Folleach replied:
resolved

Folleach commented on cs/Markdown.Tests/MdTests.cs:
```diff
@@ -0,0 +1,124 @@
+            Assert.Throws<ArgumentNullException>(() => new Md(new ParserMd(), null!));
+        }
+
+        [TestCase(100, 10, 1.65)]
+        public void Md_ShouldWorkInLinearTime(int iterations, int runsPerIteration, double threshold)
+        {
+            var md = new Md(new ParserMd(), new RendererHTML());
+
+            var averageExecutionTimes = new double[iterations];
+            var stringBuilder = new StringBuilder();
+            for (var i = 0; i < iterations; i++)
+            {
+                AddTemplates(stringBuilder);
+                var repeatedTemplates = stringBuilder.ToString();
+
+                var totalExecutionTime = 0.0;
+                for (var j = 0; j < runsPerIteration; j++)
+                {
+                    GC.Collect();
+                    GC.WaitForPendingFinalizers();
+                    var stopwatch = Stopwatch.StartNew();
+                    md.Render(repeatedTemplates);
+                    stopwatch.Stop();
+                    totalExecutionTime += stopwatch.Elapsed.TotalMilliseconds;
+                }
+                averageExecutionTimes[i] = totalExecutionTime / runsPerIteration;
+            }
+
+            for (var i = 1; i < iterations; i++)
+            {
+                var growthFactor = averageExecutionTimes[i] / averageExecutionTimes[i - 1];
+                Assert.That(growthFactor, Is.LessThanOrEqualTo(threshold));
+            }
+        }
```
У меня этот тест не проходит.

```
Assert.That(growthFactor, Is.LessThanOrEqualTo(threshold))
  Expected: less than or equal to 1.6499999999999999d
  But was:  1.8253388946819602d
```

Если я правильно понял тест, каждую итерацию к тестируемому тексту прибавляется набор `_templates`. То есть каждая итерация это `n + 1` наборов `_templates` и мы ожидаем, что прибавив один набор время не вырастет больше, чем в 1.65 раз?  
Но если у нас было уже 99 наборов в тексте, прибавляя ещё один при линейном росте время должно увеличиться в 1.01 раз?

А если у нас есть 1 набор в начале, прибавляя ещё 1 набор время должно увеличиться в 2 раза?

Вероятные причины, почему у меня тест упал:
- слишком маленькие наборы в начале, на которых рост времени велик
- нету "прогрева". При первом запуске некоторое время может тратиться на работу JIT компилятора
  https://stackoverflow.com/questions/4446203/warm-up-when-calling-methods-in-c-sharp

В дополнение, на моём компьютере после примерно 65 итерации время выполнения падает, затем резко возрастает к примерно 90 итерации.
![image](https://github.com/user-attachments/assets/e11b2401-1ef3-409c-b5ef-457f0dd229cc)
Почему? Я не знаю... Нужно рассматривать что такого происходит на этих границах.

Более концептуальный вопрос: точно ли так можно мерять сложность метода?
Кажется, прибавляя константу O(n + 1), мы должны ожидать, что разница времени будет в районе единицы для линейного роста.
В случае, если разница времени не равна единице, то мы не можем сказать, какая асимптотика у тестируемого метода.
Пример для сложности `O(n)`:
`1 + 1` - увеличение времени выполнения в 2 раза
`1000 + 1` - увеличение времени выполнения в 1.001 раз

Folleach replied:
![image](https://github.com/user-attachments/assets/0a764202-d55b-4acb-8f59-a8ef1ff5578b)

resolved

Folleach commented on cs/Markdown.Tests/MdTests.cs:
```diff
@@ -0,0 +1,124 @@
+                for (var j = 0; j < runsPerIteration; j++)
+                {
+                    GC.Collect();
+                    GC.WaitForPendingFinalizers();
```
Зачем тут вызывать GC?

Folleach replied:
resolved

BlizPerfect committed:
Измнена логика работы теста по замеру алгоритмической сложности.

1. Тест Md_ShouldWorkInLinearTime теперь работает исходя из такой логики:
Если мы увеличим длину входной строки в 2 раза, то и время выполнения работы, по сравнению с прошлой длиной входной строки, увеличится в 2 раза (+ погрешность).
2. Рефакторинг тестов.

BlizPerfect committed:
Токен теперь хранит в себе свой тип

BlizPerfect committed:
Md теперь принимает конкеретный ParserMd.

Действительно, поскольку у нас конкретный класс Md, то он должен принимать только соответсвующий себе типа парсера. К renderer это не относится.

BlizPerfect committed:
Рефакторинг кода
